pkg/reconciler/gc/gc.go

```diff
@@ -0,0 +1,151 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/serving/pkg/apis/serving"
+	"knative.dev/serving/pkg/apis/serving/v1alpha1"
+	"knative.dev/serving/pkg/apis/serving/v1beta1"
+	listers "knative.dev/serving/pkg/client/listers/serving/v1alpha1"
+	pkgreconciler "knative.dev/serving/pkg/reconciler"
+	configns "knative.dev/serving/pkg/reconciler/configuration/config"
+)
+
+// reconciler implements controller.Reconciler for Garbage Collection resources.
+type reconciler struct {
+	*pkgreconciler.Base
+
+	// listers index properties about resources
+	configurationLister listers.ConfigurationLister
+	revisionLister      listers.RevisionLister
+
+	configStore pkgreconciler.ConfigStore
+}
+
+// Check that our reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Garbage Collection
+// resource with the current status of the resource.
+func (c *reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+	logger := logging.FromContext(ctx)
+
+	ctx = c.configStore.ToContext(ctx)
+
+	// Get the Configuration resource with this namespace/name.
+	config, err := c.configurationLister.Configurations(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		// The resource no longer exists, in which case we stop processing.
+		logger.Errorf("Configuration %q in work queue no longer exists", key)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	reconcileErr := c.reconcile(ctx, config)
+	if reconcileErr != nil {
+		c.Recorder.Event(config, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+	return reconcileErr
+}
+
+func (c *reconciler) reconcile(ctx context.Context, config *v1alpha1.Configuration) error {
+	cfg := configns.FromContext(ctx).RevisionGC
+	logger := logging.FromContext(ctx)
+
+	selector := labels.Set{serving.ConfigurationLabelKey: config.Name}.AsSelector()
+	revs, err := c.revisionLister.Revisions(config.Namespace).List(selector)
+	if err != nil {
+		return err
+	}
+
+	gcSkipOffset := cfg.StaleRevisionMinimumGenerations
+
+	if gcSkipOffset >= int64(len(revs)) {
+		return nil
+	}
+
+	// Sort by creation timestamp descending
+	sort.Slice(revs, func(i, j int) bool {
+		return revs[j].CreationTimestamp.Before(&revs[i].CreationTimestamp)
+	})
+
+	for _, rev := range revs[gcSkipOffset:] {
+		if isRevisionStale(ctx, rev, config) {
+			err := c.ServingClientSet.ServingV1alpha1().Revisions(rev.Namespace).Delete(rev.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				logger.Errorf("Failed to delete stale revision %q: %v", rev.Name, err)
```

Comment:
We should be using `Errorw` whenever we report errors through our zap logger. For example,

```
logger.Errorw(fmt.Sprintf("Some error message with: %s", someValue), zap.Error(err))
```